bluefog/common/operations.cc

```diff
@@ -152,123 +138,116 @@ Response ConstructResponse(MessageTable* message_table, std::string name) {
+    tensor_shape.AddDim(dim);
+  }
+
+  if (tensor_shape.dims() == 0) {
+    error = true;
+    error_message_stream << "Rank zero tried to "
+                         << Request::RequestType_Name(message_type)
+                         << " a rank-zero tensor.";
+  }
 
-    if (tensor_shape.dims() == 0) {
+  for (unsigned int i = 1; i < requests.size(); i++) {
+    TensorShape request_shape;
+    for (auto dim : requests[i].tensor_shape()) {
+      request_shape.AddDim(dim);
+    }
+    if (tensor_shape.dims() != request_shape.dims()) {
       error = true;
-      error_message_stream << "Rank zero tried to "
+      error_message_stream << "Mismatched "
                            << Request::RequestType_Name(message_type)
-                           << " a rank-zero tensor.";
+                           << " tensor shapes: One rank sent a tensor of rank "
+                           << tensor_shape.dims()
+                           << ", but another rank sent a tensor of rank "
+                           << request_shape.dims() << ".";
+      break;
     }
 
-    for (unsigned int i = 1; i < requests.size(); i++) {
-      if (error) {
-        break;
-      }
-
-      TensorShape request_shape;
-      for (auto dim : requests[i].tensor_shape()) {
-        request_shape.AddDim(dim);
-      }
-      if (tensor_shape.dims() != request_shape.dims()) {
+    bool dim_mismatch = false;
+    for (int dim = 1; dim < tensor_shape.dims(); dim++) {
+      if (tensor_shape.dim_size(dim) != request_shape.dim_size(dim)) {
         error = true;
         error_message_stream
             << "Mismatched " << Request::RequestType_Name(message_type)
-            << " tensor shapes: One rank sent a tensor of rank "
-            << tensor_shape.dims()
-            << ", but another rank sent a tensor of rank "
-            << request_shape.dims() << ".";
-        break;
-      }
-
-      bool dim_mismatch = false;
-      for (int dim = 1; dim < tensor_shape.dims(); dim++) {
-        if (tensor_shape.dim_size(dim) != request_shape.dim_size(dim)) {
-          error = true;
-          error_message_stream
-              << "Mismatched " << Request::RequestType_Name(message_type)
-              << " tensor shapes: One rank sent a tensor with dimension " << dim
-              << " equal to " << tensor_shape.dim_size(dim)
-              << ", but another rank sent a tensor with dimension " << dim
-              << " equal to " << request_shape.dim_size(dim) << ".";
-          dim_mismatch = true;
-          break;
-        }
-      }
-      if (dim_mismatch) {
+            << " tensor shapes: One rank sent a tensor with dimension " << dim
+            << " equal to " << tensor_shape.dim_size(dim)
+            << ", but another rank sent a tensor with dimension " << dim
+            << " equal to " << request_shape.dim_size(dim) << ".";
+        dim_mismatch = true;
         break;
       }
     }
+    if (dim_mismatch) {
+      break;
+    }
   }
+  return error;
```

Comment:
return error && dim_mismatch?

Reply:
No need. The dim_mismatch is used to break the outer for loop only. When dim_mismatch is True, error should always be true. Plus, I think it should be error || dim_mismatch?